### PR TITLE
[Database] Build DatabasePersistentStorage

### LINKED ISSUE #71: [Database] Build and Test Database Production Classes
**What is the use case? Please describe.**
For this sprint task, you'll be building and testing all of the classes in the `edu.oakland.production.database` package. Note: This issue is similar to #65 , but is tracked differently as helper and production classes have different deliverable dates.

**Sprint Task Setup**

To be done by @camrongrant1 , @kylepoterek , or @MathewYaldo before the rest of the team starts working on it:

1. Edit this issue message to add a checkbox for each class, and assign it to the correct team member from #60 .

**Prerequisites**

Please makes sure that your test plan from #60 for the class you are writing is closed and merged to main before starting this, to ensure that all of the documentation you need for this is in the correct spot.

**Steps**

To be done by each team member for each class they are developing a test plan for:

1. Each team member should watch the last ~18 minutes of @andrewdimmer 's [Test Driven Development Crash Course](https://youtu.be/eSc0aY4GhPs)
2. One person should create a branch off of main called `define-Name_Of_Class` (ex. the `Demo` classes and tests get build on a branch called `build-Demo`). Everyone else in the programming pair or group should then checkout that branch.
3. Using the information in the video, use test driven development to build all the required classes, stubs, and tests. Remember: you've already defined what you need to build at https://github.com/andrewdimmer/csi-3370-software-project/tree/main/docs/uml-class-diagrams/database !
4. Commit your changes with the message `[Database] Built Name_Of_Class`
5. Push your changes
6. Create a pull request with the name `[Database] Built Name_Of_Class` and the rest of the information like description, tags, etc. For reference, the settings on the right hand side should match this issue, with the exception that the assignees on the PR are just those who are pushing code.
7. Once your code is complete, add @andrewdimmer as a reviewer.

**Tasks to Complete**

- [x] Allocate Team Members to Classes - @camrongrant1 , @kylepoterek , or @MathewYaldo
- [ ] Build DatabaseCommInterface - @Ian-Klaus 
- [x] Build DatabaseCommManager - @Endritz 
- [x] Build DatabseGisInterface - @robertsonbrinker 
- [x] Build DatabaseGisManager - @Jake-wq 
- [x] Build DatabasePersistentStorage - @MathewYaldo @letthatbe 

**Additional context**
Note: if for some reason you need to change a public interface or shared data type, reach out to @andrewdimmer who can help you manage making that BREAKING CHANGE.

Note: if you run into merge conflicts, let @andrewdimmer know and he can help you resolve them.

### PR DESCRIPTION
**Pull request by kylepoterek**
Working on #71

### CODE REVIEW

File: src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
+package edu.oakland.production.database;
+
+import edu.oakland.helper.admin.LocationDataPoint;
+import edu.oakland.helper.admin.TrackData;
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class DatabasePersistentStorageClass {
```
Comment (andrewdimmer) [inline src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java]: Based on the UML Diagrams that we created, this should be called `DatabasePersistentStorageImplementation`, and should implement the `DatabasePersistentStorage` Java interface..
```suggestion
public class DatabasePersistentStorageImplementation implements DatabasePersistentStorage {
```

File: src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+   * @param rfidData the rfid integer.
+   * @param locDataPoint the location data point array list to initialize.
+   * @param trackData the track data  array list to initialize.
+   * @param nextSatellites the string array of satellites.
+   */
+  public DatabasePersistentStorageClass(int rfidData, 
+      ArrayList<LocationDataPoint> locDataPoint, 
+      ArrayList<TrackData> trackData, 
+      String[] nextSatellites) {
+
```
Comment (andrewdimmer) [inline src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java]: I don't remember if this was in the original test spec or not, but should probably be a "Not Null" check before assigning to `locDataPoint`, `trackData` and `nextSatellites`.

File: src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+   * after providing the offset.
+   *
+   * @param offset the offset (int) of the data needed.
+   */
+
+  public LocationDataPoint getLocationDataPoint(int offset) {
+
+    // If trackdata has size of length 4 , then the offset can not be > 3 so we use size-1
+    if (offset > this.locDataPoint.size() - 1) {
+      // throw new IndexOutOfBoundsException("Offset " + offset + " is out of bounds!");
```
Comment (andrewdimmer) [inline src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java]: Just in terms of code clean up, this can be deleted.
```suggestion
```

File: src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+   *
+   * @param locationDataPoint the locationDataPoint data type.
+   */
+
+  public void storeLocationDataPoint(LocationDataPoint locationDataPoint) {
+    if (locationDataPoint == null) {
+      throw new IllegalArgumentException("locationDataPoint cannot be null");
+    }
+
+    this.locDataPoint.add(locationDataPoint);
```
Comment (andrewdimmer) [inline src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java]: This is adding to the end of the array, yet your return method is pulling from the start of the array. One of them needs to change, and it would probably be easiest if you changed this one to add to the start of the ArrayList.
```suggestion
    this.locDataPoint.add(0, locationDataPoint);
```

File: src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+   * Retrieve track data.
+   *
+   *
+   * @param offset the offset (int) of the track data needed.
+   */
+
+  public TrackData getTrackData(int offset) {
+    // If we have a td of L=4, then the offset can't be >3 so we use length-1
+    if (offset > this.trackData.size() - 1) {
+      return null;
```
Comment (andrewdimmer) [inline src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java]: Thoughts on returning an empty `TrackData`? This would greatly reduce the amount of issues Middleware_02 would have processing this output, as they always need to set the message. If might be better to do something like this (note: we would need to change the tests to make this change, but it will probable reduce an edge case later on).
```suggestion
      return new TrackData(new LocationDataPoint[0]);
```

File: src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+   *
+   * @param trackData the track data provided in the trackData data type.
+   */
+
+  public void storeTrackData(TrackData trackData) {
+    if (trackData == null) {
+      throw new IllegalArgumentException("trackData cannot be null");
+    }
+
+    this.trackData.add(trackData);
```
Comment (andrewdimmer) [inline src/main/java/edu/oakland/production/database/DatabasePersistentStorageClass.java]: This is adding to the end of the array, yet your return method is pulling from the start of the array. One of them needs to change, and it would probably be easiest if you changed this one to add to the start of the ArrayList.
```suggestion
    this.trackData.add(0, trackData);
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageStub.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package edu.oakland.test.database;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageStub.java]: Note: you don't actually need to implement or use this file in testing, but because it's already been done let's keep it if this can be merged before other the other PRs so that other people have this as a base to work off of.

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+  ArrayList<LocationDataPoint> locDataPoint = new ArrayList<LocationDataPoint>();
+  ArrayList<TrackData> trackData = new ArrayList<TrackData>();
+
+  
+
+  @Test
+  @DisplayName("RFID data in is RFID data out")
+  void rfidDataInIsRfidDataOut() {
+    //int random = (int) (Math.random() * 60);
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: I would go back to the random generation here, it makes the testing much more robust and reliable.
```suggestion
    int random = (int) (Math.random() * 60);
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    locDataPoint.add(ldp);
+    trackData.add(trackDataPoint);
+    DatabasePersistentStorageClass dpsc = new DatabasePersistentStorageClass(0, 
+        locDataPoint, 
+        trackData, 
+        randomArray
+    );
+    assertEquals(0, dpsc.locateRfidData());
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: If we're using random RFID numbers, the only change it to make this use the random number.
```suggestion
    assertEquals(random, dpsc.locateRfidData());
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+  }
+
+  @Test
+  @DisplayName("offset matches data point")
+  void offsetMatchDataPoint() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java:65]: Can we do a random offset and a random number of points for this? With only one test, this missed that we get the oldest point instead of the newest point with offset of 0 in the current implementation.

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+    );
+    int i = 0;
+    dpsc.storeLocationDataPoint(ldp);
+    assertEquals(ldp, dpsc.getLocationDataPoint(i));
+    
+  }
+
+  @Test
+  @DisplayName("Offset matches track data")
+  void offsetMatchTrackData() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java:117]: Can we do a random offset and a random number of points for this? With only one test, this missed that we get the oldest track data instead of the newest track data with offset of 0 in the current implementation.

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+
+    dpsc.storeTrackData(trackDataPoint);
+    assertEquals(trackDataPoint, dpsc.getTrackData(0));
+
+
+  }
+
+  @Test
+  @DisplayName("read Mode In Is Mode Out")
+  void readModeInIsModeOut() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: This test is redundant to the previous, so we should be able to just call it `modeInIsModeOut()` and delete the next test.
```suggestion
  void modeInIsModeOut() {
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+    trackData.add(trackDataPoint);
+    DatabasePersistentStorageClass dpsc = new DatabasePersistentStorageClass(0, 
+        locDataPoint, 
+        trackData, 
+        randomArray
+    );
+
+    dpsc.storeMode("normal");
+    assertEquals("normal", dpsc.getMode());
+  }
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: This is redundant to the above test, we can remove it.
```suggestion
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+        locDataPoint, 
+        trackData, 
+        randomArray
+    );
+    dpsc.storeLocationDataPoint(ldp);
+    assertEquals(null, dpsc.getLocationDataPoint(99999));
+  }
+
+  @Test
+  @DisplayName("track data offset out of bounds returns null")
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: Might want to update this method to just return an empty `TrackData`, see comment above. That should make it easier on Middleware.

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
Number of different comments, some stylist, some regarding functions, some regarding minor tweaks. For those that I provide a suggestion for, you can accept all the suggestions in bulk by going to https://github.com/andrewdimmer/csi-3370-software-project/pull/133/files, and adding all the suggestions to a batch before committing it at the green button at the top. For the rest that are just comments, that is something that you'll need to take a further look at. Let me know if you have any questions, and despite the number of comments this is an excellent start! Well done so far, and I look forward to seeing the final product!

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+import edu.oakland.production.database.DatabasePersistentStorageClass;
+import java.lang.IllegalArgumentException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Random;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+@DisplayName("Database PersistentStorage Test")
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: For sake of standardization to make it easier to read and search the logs, can we change this to `DatabasePersistentStorage Unit Tests`? Thank you!
```suggestion
@DisplayName("DatabasePersistentStorage Unit Tests")
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+    TrackData testTrackData = dpsc.getTrackData(99999);
+    assertNotNull(testTrackData);
+  }
+
+  @Test
+  @DisplayName("track data offset out of bounds is invalid")
+  void trackDataOffsetOutOfBoundsIsInvalid() {
+    // Made changes per r532465218
+    // Not sure if we needed to change the DisplayName and func name
+
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: We can remove this now. I like it during the change, but it just clutters up the production code afterward.
```suggestion
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 17]
```diff
@@ -0,0 +1,403 @@
...
+    locDataPoint.add(ldp);
+    trackData.add(trackDataPoint);
+    DatabasePersistentStorageImplementation dpsc = new DatabasePersistentStorageImplementation(0, 
+            locDataPoint, 
+            trackData, 
+            randomArray
+        );
+
+    // UPDATE: We return an empty trackData now.
+    //TrackData randomTd = new TrackData(new LocationDataPoint[0]);
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: We can remove this now. I like it during the change, but it just clutters up the production code afterward.
```suggestion
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 18]
```diff
@@ -0,0 +1,403 @@
...
+    dpsc.storeTrackData(trackDataPoint);
+    // test if empty track data compared to empty track data.
+    TrackData testTrackData = dpsc.getTrackData(99999);
+    assertEquals(0, testTrackData.getLocationDataPoints().length);
+  }
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: Sorry, I messed up which methods were which. This can test all 3 assert statements.
```suggestion
```

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 19]
```diff
@@ -0,0 +1,403 @@
...
+    //TrackData randomTd = new TrackData(new LocationDataPoint[0]);
+    dpsc.storeTrackData(trackDataPoint);
+    // test if empty track data compared to empty track data.
+    TrackData testTrackData = dpsc.getTrackData(99999);
+    assertNotNull(testTrackData);
+  }
+
+  @Test
+  @DisplayName("track data offset out of bounds is invalid")
+  void trackDataOffsetOutOfBoundsIsInvalid() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: Sorry, I messed up which methods were which. This can test all 3 assert statements.
```suggestion
  @DisplayName("track data offset out of bounds returns empty track data")
  void trackDataOffsetOutOfBoundsReturnsEmptyTrackData() {
```

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
More more minor group of changes that can be taken in a batch, then you're good to go!

File: src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java [hunk 20]
```diff
@@ -0,0 +1,452 @@
...
+            randomArray
+        );
+
+    // UPDATE: We return an empty trackData now.
+    //TrackData randomTd = new TrackData(new LocationDataPoint[0]);
+    dpsc.storeTrackData(trackDataPoint);
+    // test if empty track data compared to empty track data.
+    TrackData testTrackData = dpsc.getTrackData(99999);
+    assertNotNull(testTrackData);
+  }
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabasePersistentStorageTest.java]: Sorry, I messed up which methods were which. This can test all 3 assert statements.
```suggestion
```

**Review by MathewYaldo — APPROVED** [review]

**Review by andrewdimmer — APPROVED** [review]
Looks good, well done! You're all set to merge!

**Review by camrongrant1 — APPROVED** [review]
